baby-liminal-extension/src/api.rs

```diff
@@ -0,0 +1,104 @@
+use ink::{
+    env::{DefaultEnvironment, Environment as EnvironmentT},
+    prelude::vec::Vec,
+    primitives::AccountId,
+};
+
+use crate::VerificationKeyIdentifier;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, scale::Encode, scale::Decode)]
+#[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+/// Chain extension errors enumeration.
+pub enum BabyLiminalError {
+    // `pallet_baby_liminal::store_key` errors
+    VerificationKeyTooLong,
+    IdentifierAlreadyInUse,
+    StoreKeyErrorUnknown,
+
+    // `pallet_baby_liminal::verify` errors
+    UnknownVerificationKeyIdentifier,
+    DeserializingProofFailed,
+    DeserializingPublicInputFailed,
+    DeserializingVerificationKeyFailed,
+    VerificationFailed,
+    IncorrectProof,
+    VerifyErrorUnknown,
+
+    ScaleError,
+    UnknownError,
+}
+
+impl From<scale::Error> for BabyLiminalError {
+    fn from(_: scale::Error) -> Self {
+        Self::ScaleError
+    }
+}
+
+impl ink::env::chain_extension::FromStatusCode for BabyLiminalError {
+    fn from_status_code(status_code: u32) -> Result<(), Self> {
+        use crate::error_codes::*;
+
+        match status_code {
+            // Success codes
+            BABY_LIMINAL_STORE_KEY_SUCCESS | BABY_LIMINAL_VERIFY_SUCCESS => Ok(()),
+
+            // `pallet_baby_liminal::store_key` errors
+            BABY_LIMINAL_STORE_KEY_TOO_LONG_KEY => Err(Self::VerificationKeyTooLong),
+            BABY_LIMINAL_STORE_KEY_IDENTIFIER_IN_USE => Err(Self::IdentifierAlreadyInUse),
+            BABY_LIMINAL_STORE_KEY_ERROR_UNKNOWN => Err(Self::StoreKeyErrorUnknown),
+
+            // `pallet_baby_liminal::verify` errors
+            BABY_LIMINAL_VERIFY_DESERIALIZING_PROOF_FAIL => Err(Self::DeserializingProofFailed),
+            BABY_LIMINAL_VERIFY_DESERIALIZING_INPUT_FAIL => {
+                Err(Self::DeserializingPublicInputFailed)
+            }
+            BABY_LIMINAL_VERIFY_UNKNOWN_IDENTIFIER => Err(Self::UnknownVerificationKeyIdentifier),
+            BABY_LIMINAL_VERIFY_DESERIALIZING_KEY_FAIL => {
+                Err(Self::DeserializingVerificationKeyFailed)
+            }
+            BABY_LIMINAL_VERIFY_VERIFICATION_FAIL => Err(Self::VerificationFailed),
+            BABY_LIMINAL_VERIFY_INCORRECT_PROOF => Err(Self::IncorrectProof),
+            BABY_LIMINAL_VERIFY_ERROR_UNKNOWN => Err(Self::VerifyErrorUnknown),
+
+            _ => Err(Self::UnknownError),
+        }
+    }
+}
+
+/// BabyLiminal chain extension definition.
+#[ink::chain_extension]
+pub trait BabyLiminalExtension {
+    type ErrorCode = BabyLiminalError;
+
+    /// Directly call `pallet_baby_liminal::store_key`.
+    #[ink(extension = 41)]
+    fn store_key(
+        origin: AccountId,
+        identifier: VerificationKeyIdentifier,
+        key: Vec<u8>,
+    ) -> Result<(), BabyLiminalError>;
+
+    /// Directly call `pallet_baby_liminal::verify`.
+    #[ink(extension = 42)]
```

Comment:
Can we use `extension_ids` here? Or does the macro won't work with variables?

Reply:
please take a look at the current code version (I added note in the latest commit)

Reply:
Yeah, I just saw it https://github.com/Cardinal-Cryptography/aleph-node/pull/1498/commits/676e9ff7f2f1be8c522dbfaaa269e55cdb70cca7.